### PR TITLE
Update cosmos

### PR DESCRIPTION
**Pull request by jsancio**
High level description including:
- What does this change enable
  - Update Cosmos to support the following new endpoints
    - /package/add - Add a package to the local repository
    - /service/start - Starts a service using Marathon for the given package
  - Change the configuration generator to allow the user to optionally provide permanent store for packages. If the user configure Cosmos with this information it enables the features outlined above.
- What bugs does this change fix
  - None
- High-Level how things changed
  - See above.

# Issues

[DCOS-8992](https://dcosjira.atlassian.net/browse/DCOS-8992)

# Checklist

 - [x] There are extensive unit and CI tests in the Cosmos repo.
 - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

If this is a component/package update, include
 - This feature is integrated tested in the Cosmos CI builds. See [results](https://teamcity.mesosphere.io/viewType.html?buildTypeId=DcosIo_Cosmos_Ci).

### CODE REVIEW

File: packages/cosmos/build [hunk 1]
```diff
@@ -9,6 +9,8 @@ chmod -R o+r "${PKG_PATH}/usr"
 cosmos_service="${PKG_PATH}/dcos.target.wants_master/dcos-cosmos.service"
 mkdir -p $(dirname "$cosmos_service")
 
+source /opt/mesosphere/etc/cosmos.env
```
Comment (jsancio) [inline packages/cosmos/build]: @cmaloney Is this valid?

File: gen/calc.py [hunk 2]
```diff
@@ -421,6 +421,64 @@ def calculate_cluster_docker_registry_enabled(cluster_docker_registry_url):
     return 'false' if cluster_docker_registry_url == '' else 'true'
 
 
+def validate_cosmos_config(cosmos_config):
```
Comment (jsancio) [inline gen/calc.py]: What is `cosmos_config`? I am sure how this symbol is getting resolved. I notice an increasing correlation between the name of the parameter and the name of the property in `entry`. Is this how it is resolved?

File: packages/cosmos/build (hunk 1 above)
Comment (spahl) [inline packages/cosmos/build]: No you don't have /opt/mesosphere/ in the build environment. None of these files are generated or rendered in there.

Comment (jsancio) [inline packages/cosmos/build]: I am open to suggestions around this. Do I source the file in the actual unit and replace `$COSMOS...` with `\$COSMOS...`?

Comment (spahl) [inline packages/cosmos/build]: Correct. You source the file with systemd and escape.

Comment (jsancio) [inline packages/cosmos/build]: Done. Thanks!

File: gen/calc.py (hunk 2 above)
Comment (branden) [inline gen/calc.py]: Yeah, we use `inspect` to find the names of this validator's arguments, and bind them to the values of the config parameters with those same names. https://github.com/dcos/dcos/blob/master/gen/internals.py#L334

Comment (branden) [inline gen/calc.py]: Same story for calculate functions, btw. https://github.com/dcos/dcos/blob/master/gen/internals.py#L61

**Comment by jsancio** [comment]
@branden and @spahl. This is ready for review and merging.

File: gen/calc.py [hunk 3]
```diff
@@ -420,6 +420,64 @@ def calculate_cluster_docker_registry_enabled(cluster_docker_registry_url):
...
+          "type": "string"
+        },
+        "package_storage_uri": {
+          "type": "string"
+        }
+      }
+    }
+    """
+
+    config = json.loads(cosmos_config)
```
Comment (cmaloney) [inline gen/calc.py]: Please use https://github.com/dcos/dcos/blob/master/gen/calc.py#L187 to load out this dictionary, makes nice errors if the user provides something not json

File: gen/calc.py [hunk 4]
```diff
@@ -420,6 +420,64 @@ def calculate_cluster_docker_registry_enabled(cluster_docker_registry_url):
...
+    if ('staged_package_storage_uri' in config and
```
Comment (cmaloney) [inline gen/calc.py]: These get turned into environment variables. What happens if a customer has a space inside their storage uri?

File: gen/calc.py (hunk 3 above)
Comment (jsancio) [inline gen/calc.py]: Done.

File: gen/calc.py (hunk 4 above)
Comment (jsancio) [inline gen/calc.py]: Done. I am validating that it is an URL. That should also validate that it doesn't contain spaces.

**Comment by jsancio** [comment]
@cmaloney Thanks for the review. I addressed your inline comments. Working on getting tests going for the additional configuration option.

**Comment by cmaloney** [comment]
The vagrant/bash failures being seen earlier (And I'm guessing which will reoccur with the latest push) are actual failures caused by changes in this pr https://teamcity.mesosphere.io/viewLog.html?buildId=500151&tab=buildResultsDiv&buildTypeId=DcosIo_Dcos_VagrantIntegrationTests_ProvisioningTestGenconfBashPr

All the journal logs are available for the test runs in the "Cleanup" step: https://teamcity.mesosphere.io/viewLog.html?buildId=500151&buildTypeId=DcosIo_Dcos_VagrantIntegrationTests_ProvisioningTestGenconfBashPr&tab=buildLog#_state=10550,10556&focus=10556

**Comment by jsancio** [comment]
@cmaloney Are you saying that this PR is causing the vagrant tests to fail?

**Comment by cmaloney** [comment]
Yes, when the vagrant integration test runs it fails because of changes this PR makes.

**Comment by jsancio** [comment]
I found this:
```
[21:35:44][m3] Dec 06 13:33:17 m3.dcos java[7799]: Processed session termination for sessionid: 0x158d6093efc0079
[21:35:44][m3] Dec 06 13:33:17 m3.dcos java[13659]: Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x00000000d14a0000, 320180224, 0) failed; error='Cannot allocate memory' (errno=12)
[21:35:44][m3] Dec 06 13:33:17 m3.dcos java[13659]: #
[21:35:44][m3] Dec 06 13:33:17 m3.dcos java[13659]: # There is insufficient memory for the Java Runtime Environment to continue.
[21:35:44][m3] Dec 06 13:33:17 m3.dcos java[13659]: # Native memory allocation (mmap) failed to map 320180224 bytes for committing reserved memory.
[21:35:44][m3] Dec 06 13:33:17 m3.dcos java[13659]: # An error report file with more information is saved as:
[21:35:44][m3] Dec 06 13:33:17 m3.dcos java[13659]: # /tmp/hs_err_pid13659.log
[21:35:44][m3] Dec 06 13:33:17 m3.dcos systemd[1]: dcos-cosmos.service: main process exited, code=exited, status=1/FAILURE
[21:35:44][m3] Dec 06 13:33:18 m3.dcos java[7799]: Expiring session 0x158d6093efc0073, timeout of 10000ms exceeded
[21:35:44][m3] Dec 06 13:33:18 m3.dcos java[7799]: Processed session termination for sessionid: 0x158d6093efc0073
[21:35:44][m3] Dec 06 13:33:18 m3.dcos systemd[1]: Unit dcos-cosmos.service entered failed state.
[21:35:44][m3] Dec 06 13:33:18 m3.dcos systemd[1]: dcos-cosmos.service failed.
```

**Comment by cmaloney** [comment]
Another high level comment: How does this expect to get permissions / credentials to push stuff to the storage uri when a package is being added?

**Comment by jsancio** [comment]
> Another high level comment: How does this expect to get permissions / credentials to push stuff to the storage uri when a package is being added?

We support two object store:

1. If the object store is the file system then the credential will come from the process.
2. If the object store is S3 then we use the standard aws chain for getting credentials: env, java system property, etc. http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html. When we move to providing a secret store we can design a different solution and maybe expose it in the gen config.

For now we are going to recommend users to either use the file system object store or S3 with EC2 roles.

**Comment by cmaloney** [comment]
The standard AWS lookup scheme isn't functional / desired for s3 based on experience accessing s3 for exhibitor storage.

For s3 users need to be able to:

1. in config.yaml provide explicit credentials. The IAM credentials of a host often don't match the credentials needed.
2. in config.yaml specify to load the credentials out of some file in `/etc` that is managed by their credential management system
3. Defer to IAM (Which should be the config for our AWS CF templates)

They have no way of injecting environment variables at runtime into our services (And that is by design). They also can't edit the java system properties.

When you say for filesystem the credentials come from the process do you mean "credentials are just the standard filesystem permissions of the user cosmos runs as"?

**Comment by jsancio** [comment]
@cmaloney The tests pass. The issue was that the Vagrant VM doesn't have much memory and the JVM was running out of memory loading the Jar. I took a look at the Jar and we were including Jars from the AWS SDK that we didn't need. I removed those dependency, now the Jar is small enough for Vagrant. We should look into changing that integration test and increasing the VM memory. cc @karlkfi 

Regarding your credential issue. This is an experimental issue. We will document it as only supporting the file system option. Before we remove the experimental tag we will add support for S3 credential in the `config.yaml`. What do you think? cc @cruhland

**Comment by karlkfi** [comment]
VM mem is entirely up to the dcos-vagrant user. It's configurable in the VagrantConfig.yaml. 

-Karl

**Comment by cmaloney** [comment]
If we've started using quite a bit more memory than we used to, that likely means we need to up the minimum requirements of DC/OS. Currently we use the default VagrantConfig.yaml, and if that can't run DC/OS it means we probably need to update the default to have more memory on the masters.

**Comment by karlkfi** [comment]
The default mem on the master is pretty low (1GB). I can up the default if you tell me what it actually needs. It was already paging rather aggressively at 1GB, but any change needs to be balanced against a reduced space for deploying apps on a standard 16GB dev laptop. So we can't just up it arbitrarily high.

**Comment by jsancio** [comment]
Back to you @cmaloney and @mellenburg. Ready for review. The failures I am seeing seem to be flaky tests, please confirm.

**Comment by jsancio** [comment]
@mellenburg I used the following value in `config.yaml` during my manual tests:
```
cosmos_config:
  staged_package_storage_uri: file:///var/lib/dcos/cosmos/staged-packages
  package_storage_uri: file:///var/lib/dcos/cosmos/packages
```

**Comment by jsancio** [comment]
Success! All of the tests pass. cc @lingmann

File: gen/calc.py [hunk 5]
```diff
@@ -419,6 +429,68 @@ def calculate_cluster_docker_registry_enabled(cluster_docker_registry_url):
...
+          "type": "string"
+        },
+        "package_storage_uri": {
+          "type": "string"
+        }
+      }
+    }
+    """
+    config = validate_json_dictionary(cosmos_config)
+    if ('staged_package_storage_uri' in config and
```
Comment (cmaloney) [inline gen/calc.py]: How about something like the following? I think it simplifies the code a bit, makes it harder to get wrong.

```
expected = ['staged_package_storage_uri', 'package_storage_uri']
found = list(filter(lambda param: param in config, expected))
if len(found) == 0:
    return

if len(found) == 1:
    raise AssertionError(
        'cosmos_config must be a dictionary containing both {}, or must be left '
        'empty. Found only {}'.format(' '.join(expected), found))

for parameter in expected:
    validate_url(config[parameter])
```

File: gen/calc.py [hunk 6]
```diff
@@ -419,6 +429,68 @@ def calculate_cluster_docker_registry_enabled(cluster_docker_registry_url):
...
+        )
+    elif ('staged_package_storage_uri' in config and
+          'package_storage_uri' in config):
+        # Let's try to parse the string as an URL
+        validate_url(config['staged_package_storage_uri'])
+        validate_url(config['package_storage_uri'])
+
+
+def calculate_cosmos_staged_package_storage_uri_flag(cosmos_config):
+    if 'staged_package_storage_uri' in cosmos_config:
```
Comment (cmaloney) [inline gen/calc.py]: this is doing string searching rather than dictionary lookup. Can we switch this around to

1. load_json(cosmos_config)
2. check if key exists and if so return
3. else return empty

File: gen/calc.py [hunk 7]
```diff
@@ -419,6 +429,68 @@ def calculate_cluster_docker_registry_enabled(cluster_docker_registry_url):
...
+            '-com.mesosphere.cosmos.stagedPackageStorageUri={}'.format(
+                json.loads(cosmos_config)['staged_package_storage_uri']
+            )
+        )
+    else:
+        return ''
+
+
+def calculate_cosmos_package_storage_uri_flag(cosmos_config):
+    if 'package_storage_uri' in cosmos_config:
```
Comment (cmaloney) [inline gen/calc.py]: Same as above

File: packages/adminrouter/buildinfo.json [hunk 8]
```diff
@@ -9,7 +9,7 @@
       "adminrouter": {
           "kind": "git",
           "git": "https://github.com/dcos/adminrouter.git",
-          "ref": "5e23b06be6c061b394d2123424675c542072a6f2",
+          "ref": "44bb0b2d119154f9b4ea9f1c3af184c278bb39b5",
```
Comment (cmaloney) [inline packages/adminrouter/buildinfo.json]: nit: this is causing adminrouter to include a lot of unrelated changes which aren't tested or talked about in this PR.

File: packages/dcos-integration-test/extra/test_endpoints.py [hunk 9]
```diff
@@ -144,6 +146,50 @@ def test_if_we_have_capabilities(cluster):
     assert {'name': 'PACKAGE_MANAGEMENT'} in r.json()['capabilities']
 
 
+def test_cosmos_package_add(cluster):
```
Comment (cmaloney) [inline packages/dcos-integration-test/extra/test_endpoints.py]: @mellenburg is going to need to review these tests. I'm doing a first pass here.

File: packages/dcos-integration-test/extra/test_endpoints.py [hunk 10]
```diff
@@ -144,6 +146,50 @@ def test_if_we_have_capabilities(cluster):
...
+    def package_add_configured():
+        user_config = load_yaml("/opt/mesosphere/etc/user.config.yaml")
```
Comment (cmaloney) [inline packages/dcos-integration-test/extra/test_endpoints.py]: The user config isn't what we want to load here. Ideally we want to say from outside the cluster that is running what state the cluster should be in for these conditional things. (See things like dns_search in https://github.com/dcos/dcos/blob/master/test_util/runner.py#L17). 

As a midway point, can for now probably load `/opt/mesosphere/etc/expanded.config.json` which has all the configuration values which were set or calculated. and grab the flags out of there. There might be a helper around for getting that file, talk to @mellenburg

File: packages/dcos-integration-test/extra/test_endpoints.py [hunk 11]
```diff
@@ -144,6 +146,50 @@ def test_if_we_have_capabilities(cluster):
...
+        json={
+            'packageName': 'cassandra',
+            'packageVersion': '1.0.20-3.0.10'
+        }
+    )
+
+    if package_add_configured():
+        # if the config is enabled then Cosmos should accept the request and
+        # return 202
+        assert r.status_code == 202
```
Comment (cmaloney) [inline packages/dcos-integration-test/extra/test_endpoints.py]: `assert r.status_code == 202` ends up producing really hard to debug / track down errors when it fails (you don't get the status code or response body printed which are both usually critical in solving). Please use r.raise_for_status() or do a `assert r.status_code == 202, "status_code: {}, content: {}".format(r.status_code, r.content))`

File: gen/calc.py (hunk 5 above)
Comment (jsancio) [inline gen/calc.py]: Fixed. Changed the algorithm above slightly.

File: gen/calc.py (hunk 6 above)
Comment (jsancio) [inline gen/calc.py]: Oh, wow. Good catch. Fixed.

File: gen/calc.py (hunk 7 above)
Comment (jsancio) [inline gen/calc.py]: Fixed.

File: packages/adminrouter/buildinfo.json (hunk 8 above)
Comment (jsancio) [inline packages/adminrouter/buildinfo.json]: I am open to suggestions here, I am not sure how to fix it. We need the changes in that commit to expose Cosmos correctly. Do you want me to change the AdminRouter in a different PR?

File: packages/dcos-integration-test/extra/test_endpoints.py (hunk 10 above)
Comment (jsancio) [inline packages/dcos-integration-test/extra/test_endpoints.py]: I fixed it to load the JSON you enumerated above. I greped the code and didn't find an util function. @mellenburg Let know if I missed it.

File: packages/dcos-integration-test/extra/test_endpoints.py (hunk 11 above)
Comment (jsancio) [inline packages/dcos-integration-test/extra/test_endpoints.py]: Good point. Fixed.

**Comment by jsancio** [comment]
Thanks for the review @cmaloney!

@mellenburg, Let @cmaloney and I know if there are any changes you would like me to make.

**Comment by jsancio** [comment]
> dcos-docs PR for adding the new configuration options (https://github.com/dcos/dcos-docs/blob/master/1.9/administration/installing/custom/configuration-parameters.md)

I am working on this. I was going to do this after this PR gets merged but I can do this now.

**Comment by jsancio** [comment]
> dcos-docs PR for adding the new configuration options (https://github.com/dcos/dcos-docs/blob/master/1.9/administration/installing/custom/configuration-parameters.md)

https://github.com/dcos/dcos-docs/pull/787

**Comment by mellenburg** [comment]
@jsancio Here is a job that is running with the additional config injected: https://teamcity.mesosphere.io/viewLog.html?buildId=504633&buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_DcosOssCcmIntegration_2&tab=buildLog

This can be toggled by setting the env var ENABLE_COSMOS_ADVANCED_TESTING in onprem TeamCity integration tests (set to default of false until this PR is merged, then will be enabled on all PRs). You can force this setting by running a custom build on this branch and filling in true in the parameters section

File: packages/dcos-integration-test/extra/test_endpoints.py [hunk 12]
```diff
@@ -144,6 +146,49 @@ def test_if_we_have_capabilities(cluster):
     assert {'name': 'PACKAGE_MANAGEMENT'} in r.json()['capabilities']
 
 
+def test_cosmos_package_add(cluster):
+    def package_add_configured():
+        user_config = load_json("/opt/mesosphere/etc/expanded.config.json")
```
Comment (mellenburg) [inline packages/dcos-integration-test/extra/test_endpoints.py]: Please only use inner functions if they must take advantage of local variables. There are no such variables being used here

File: packages/dcos-integration-test/extra/test_endpoints.py [hunk 13]
```diff
@@ -144,6 +146,49 @@ def test_if_we_have_capabilities(cluster):
...
+                'charset=utf-8;version=v1'
+            )
+        },
+        json={
+            'packageName': 'cassandra',
+            'packageVersion': '1.0.20-3.0.10'
+        }
+    )
+
+    if package_add_configured():
```
Comment (mellenburg) [inline packages/dcos-integration-test/extra/test_endpoints.py]: This function can be removed outright in favor of:
```
user_config = load_json("/opt/mesosphere/etc/expanded.config.json")
if user_config['cosmos_staged_package_storage_uri_flag'] and user_config['cosmos_package_storage_uri_flag']:
     # same code
```

File: packages/dcos-integration-test/extra/test_endpoints.py (hunk 12 above)
Comment (jsancio) [inline packages/dcos-integration-test/extra/test_endpoints.py]: Removed.

File: packages/dcos-integration-test/extra/test_endpoints.py (hunk 13 above)
Comment (jsancio) [inline packages/dcos-integration-test/extra/test_endpoints.py]: Done.

File: packages/dcos-integration-test/extra/test_endpoints.py [hunk 14]
```diff
@@ -144,6 +146,43 @@ def test_if_we_have_capabilities(cluster):
...
+            'packageVersion': '1.0.20-3.0.10'
+        }
+    )
+
+    user_config = load_json("/opt/mesosphere/etc/expanded.config.json")
+    if (user_config['cosmos_staged_package_storage_uri_flag'] and
+            user_config['cosmos_package_storage_uri_flag']):
+        # if the config is enabled then Cosmos should accept the request and
+        # return 202
+        assert r.status_code == 202, 'status = {}, content = {}'.format(
```
Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_endpoints.py]: Minor comment.

These assert lines could be folded into a single line as the fall under our 120 col limit.

File: packages/adminrouter/buildinfo.json (hunk 8 above)
Comment (adam-mesos) [inline packages/adminrouter/buildinfo.json]: Adminrouter has been bumped. Now this bump only includes 3 commits to AR: debugging-support attempt1 (accidentally reverted by unified-proxy), unified-proxy, and cosmos-endpoint.
Needs a rebase, but you can keep the 44bb SHA
